dandischema/digests/zarr.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+import hashlib
+from typing import Dict, List, Optional
+
+import pydantic
+
+
+"""Passed to the json() method of pydantic models for serialization."""
+ENCODING_KWARGS = {"separators": (",", ":")}
+
+
+class ZarrChecksum(pydantic.BaseModel):
+    """
+    A checksum for a single file/directory in a zarr file.
+
+    Every file and directory in a zarr archive has a path and a MD5 hash.
+    """
+
+    md5: str
+    path: str
+
+    # To make ZarrChecksums sortable
+    def __lt__(self, other: ZarrChecksum):
+        return self.path < other.path
+
+
+class ZarrChecksums(pydantic.BaseModel):
+    """
+    A set of file and directory checksums.
+
+    This is the data hashed to calculate the checksum of a directory.
+    """
+
+    directories: List[ZarrChecksum] = pydantic.Field(default_factory=list)
+    files: List[ZarrChecksum] = pydantic.Field(default_factory=list)
+
+    @property
+    def is_empty(self):
+        return self.files == [] and self.directories == []
+
+    def _index(self, checksums: List[ZarrChecksum], checksum: ZarrChecksum):
+        # O(n) performance, consider an ordered dict for optimization
+        for i in range(0, len(checksums)):
+            if checksums[i].path == checksum.path:
+                return i
+        raise ValueError("Not found")
+
+    def add_file_checksums(self, checksums: List[ZarrChecksum]):
+        for new_checksum in checksums:
+            try:
+                self.files[self._index(self.files, new_checksum)] = new_checksum
+            except ValueError:
+                self.files.append(new_checksum)
+        self.files = sorted(self.files)
+
+    def add_directory_checksums(self, checksums: List[ZarrChecksum]):
+        """Add a list of directory checksums to the listing."""
+        for new_checksum in checksums:
+            try:
+                self.directories[
+                    self._index(self.directories, new_checksum)
+                ] = new_checksum
+            except ValueError:
+                self.directories.append(new_checksum)
+        self.directories = sorted(self.directories)
+
+    def remove_checksums(self, paths: List[str]):
+        """Remove a list of paths from the listing."""
+        self.files = sorted(
+            filter(lambda checksum: checksum.path not in paths, self.files)
+        )
+        self.directories = sorted(
```

Comment:
didn't check in details what tests do, but it seems that all above functionality in ZarrChecksums is not tested at all! Could some basic test(s) be added to help ensure no surprises? (ideally on some case where there is a known target checksum(s))

Reply:
Added some tests, `zarr.py` now has 100% coverage